### PR TITLE
Generic errors in proto

### PR DESCRIPTION
**Pull request by bluejekyll**
This should resolve the issue with the `NXDOMAIN` detection. We can't do anything with the trustworthiness of the answers from upstream resolvers and authorities, but we can at least surface the correct error.

Fixes: #1171

### CODE REVIEW

**Comment by glts** [comment]
@bluejekyll Thank you! Looks like it’s working with my example program (in the link, the [workaround](https://gitlab.com/glts/viaspf/-/blob/500d2b8f0607a989d27d03c4e6bd914f3371dc92/examples/spfquery.rs#L78-82) seems no longer necessary).

Can you say if your fix will encode both NXDOMAIN and NOERROR with empty answer as `ResolveErrorKind::NoRecordsFound`?

**Comment by bluejekyll** [comment]
Yes, it should: https://github.com/bluejekyll/trust-dns/blob/70360e09abcf2f41e854bf9885ae39ee7b1d1916/crates/resolver/src/name_server/name_server.rs#L183

We might be able to get a better TTL there, I’m going to look into that...

File: crates/resolver/src/name_server/connection_provider.rs [hunk 1]
```diff
@@ -88,7 +90,7 @@ pub trait RuntimeProvider: Clone + 'static {
 pub trait Spawn {
     fn spawn_bg<F>(&mut self, future: F)
     where
-        F: Future<Output = Result<(), ProtoError>> + Send + 'static;
+        F: Future<Output = Result<(), ResolveError>> + Send + 'static;
```
Comment (bluejekyll) [inline crates/resolver/src/name_server/connection_provider.rs:93]: @chunyingw, you might be interested in this specific change

File: crates/resolver/src/name_server/name_server.rs [hunk 2]
```diff
@@ -143,21 +149,40 @@ impl<C: DnsHandle, P: ConnectionProvider<Conn = C>> NameServer<C, P> {
...
+                                query,
+                                valid_until: None,
+                                response_code: ResponseCode::NXDomain,
+                            };
+                            return Err(ResolveError::from(error_kind));
                         }
                         ResponseCode::NoError
                             if response.answers().is_empty()
                                 && response.additionals().is_empty()
                                 && response.name_servers().is_empty() =>
```
Comment (bluejekyll) [inline crates/resolver/src/name_server/name_server.rs:172]: we only care about name_servers in the case where we are looking up specific record types, NS, PTR, NAPTR. We may need a specific check for this?

File: crates/resolver/src/name_server/name_server.rs [hunk 3]
```diff
@@ -143,21 +149,40 @@ impl<C: DnsHandle, P: ConnectionProvider<Conn = C>> NameServer<C, P> {
...
                             let note = "Nameserver responded with NXDomain";
                             debug!("{}", note);
-                            return Err(ProtoError::from(note));
+
+                            let mut response = response;
+                            let query =
+                                response.take_queries().drain(..).next().unwrap_or_default();
+                            let error_kind = ResolveErrorKind::NoRecordsFound {
+                                query,
+                                valid_until: None,
```
Comment (bluejekyll) [inline crates/resolver/src/name_server/name_server.rs:164]: check for an SOA record in the response, name_servers section, that would have the TTL for this value.

**Comment by bluejekyll** [comment]
@glts I added some new code to more accurately determine the NOERROR no data response. This should be a big improvement for your use case. This adds checks that conform with https://tools.ietf.org/html/rfc2308#section-2

**Comment by bluejekyll** [comment]
This is pretty close to complete. I just want to evaluate caching to make sure that is correct.

**Pull request merged**